distributed-counter/src/main/java/org/infinispan/counter/impl/function/ResetFunction.java

```diff
@@ -0,0 +1,73 @@
+package org.infinispan.counter.impl.function;
+
+import org.infinispan.commons.api.functional.EntryView;
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.counter.impl.entries.CounterValue;
+import org.infinispan.counter.impl.externalizers.ExternalizerIds;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * A function to reset a {@link CounterValue}.
+ * <p>
+ * This function is a singleton.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+public class ResetFunction<T> implements Function<EntryView.ReadWriteEntryView<T, CounterValue>, Void> {
+
+
+   private ResetFunction() {
+   }
+
+   public static <T> ResetFunction<T> getInstance() {
+      //noinspection unchecked
+      return SingletonHolder.INSTANCE;
+   }
+
+   @Override
+   public Void apply(EntryView.ReadWriteEntryView<T, CounterValue> entryView) {
+      Optional<CounterValue> currentValue = entryView.find();
+      if (currentValue.isPresent()) {
+         CounterValue currentEntry = currentValue.get();
+         CounterValue newEntry = currentEntry.reset();
+         if (newEntry != null) {
+            entryView.set(newEntry);
+         }
+      }
+      return null;
+   }
+
+   private static class SingletonHolder {
+      private static final ResetFunction INSTANCE = new ResetFunction();
```

Comment:
This pattern should be used when the instance is heavy and should be instantiated lazily on getInstance() call... that's a bit of overkill here, I'd move the instance to the main class.